# Real Estate Price Prediction

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
import matplotlib
matplotlib.rcParams['figure.figsize'] = (20,10)

In [2]:
df = pd.read_csv(r"C:\BrushUp\PythonBasics\Project\Real_Estate\bengaluru_house_prices.csv")
df

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK      NaN       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK      NaN       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK      NaN       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [3]:
df.shape

(13320, 9)

In [4]:
df.groupby("area_type")["area_type"].agg('count')

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

In [5]:
#Removing Unwanted Column
df2 = df.drop(['area_type', 'society', 'balcony','availability'], axis='columns')
df2

location       size total_sqft  bath   price
0      Electronic City Phase II      2 BHK       1056   2.0   39.07
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2                   Uttarahalli      3 BHK       1440   2.0   62.00
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                      Kothanur      2 BHK       1200   2.0   51.00
...                         ...        ...        ...   ...     ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00
13316             Richards Town      4 BHK       3600   5.0  400.00
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00
13319              Doddathoguru      1 BHK        550   1.0   17.00

[13320 rows x 5 columns]

### Data Cleaning

In [6]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [7]:
df3 = df2.dropna()
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [8]:
df3['size']

0            2 BHK
1        4 Bedroom
2            3 BHK
3            3 BHK
4            2 BHK
           ...    
13315    5 Bedroom
13316        4 BHK
13317        2 BHK
13318        4 BHK
13319        1 BHK
Name: size, Length: 13246, dtype: object

In [9]:
df3['bhk'] = df3['size'].apply(lambda x: int(x.split(" ")[0]))
df3['bhk']

C:\Users\devadath.nair\AppData\Local\Temp\ipykernel_33340\3966064352.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['bhk'] = df3['size'].apply(lambda x: int(x.split(" ")[0]))


0        2
1        4
2        3
3        3
4        2
        ..
13315    5
13316    4
13317    2
13318    4
13319    1
Name: bhk, Length: 13246, dtype: int64

In [10]:
df3.head()

location       size total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK       1056   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00    4
2               Uttarahalli      3 BHK       1440   2.0   62.00    3
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00    3
4                  Kothanur      2 BHK       1200   2.0   51.00    2

In [11]:
df3['bhk'].unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18], dtype=int64)

In [12]:
df3[df3.bhk>10]

location        size total_sqft  bath  price  bhk
459                     1 Giri Nagar      11 BHK       5000   9.0  360.0   11
1718       2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
1768              1 Ramamurthy Nagar  11 Bedroom       1200  11.0  170.0   11
3379                  1Hanuman Nagar      19 BHK       2000  16.0  490.0   19
3609   Koramangala Industrial Layout      16 BHK      10000  16.0  550.0   16
3853               1 Annasandrapalya  11 Bedroom       1200   6.0  150.0   11
4684                     Munnekollal  43 Bedroom       2400  40.0  660.0   43
4916                   1Channasandra      14 BHK       1250  15.0  125.0   14
6533                     Mysore Road  12 Bedroom       2232   6.0  300.0   12
7979                   1 Immadihalli      11 BHK       6000  12.0  150.0   11
9935                   1Hoysalanagar      13 BHK       5425  13.0  275.0   13
11559                  1Kasavanhalli  18 Bedroom       1200  18.0  200.0   18

In [13]:
df3['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [14]:
def is_floata(x):
    try:
        float(x)
    except:
        return False
    return True

In [15]:
df3[~df3['total_sqft'].apply(is_floata)].head(10)

location       size      total_sqft  bath    price  bhk
30            Yelahanka      4 BHK     2100 - 2850   4.0  186.000    4
122              Hebbal      4 BHK     3067 - 8156   4.0  477.000    4
137  8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005    2
165            Sarjapur      2 BHK     1145 - 1340   2.0   43.490    2
188            KR Puram      2 BHK     1015 - 1540   2.0   56.800    2
410             Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500    1
549         Hennur Road      2 BHK     1195 - 1440   2.0   63.770    2
648             Arekere  9 Bedroom       4125Perch   9.0  265.000    9
661           Yelahanka      2 BHK     1120 - 1145   2.0   48.130    2
672        Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000    4

In [16]:
a= "2100 - 2850"
a.split(" - ")

['2100', '2850']

In [17]:
def cnvrt_sqft_to_num(x):
    tokens = x.split(" - ")
    if len(tokens)==2:
        return(float(tokens[0])+float(tokens[1]))/2
    try:
        return float(x)
    except:
        return None
        

In [18]:
cnvrt_sqft_to_num('2100')

2100.0

In [19]:
cnvrt_sqft_to_num("2100 - 2850")

2475.0

In [20]:
cnvrt_sqft_to_num("34.46Sq. Meter")

In [21]:
df4 = df3.copy()
df4["total_sqft"] = df4["total_sqft"].apply(cnvrt_sqft_to_num)
df4.head(31)

location       size  total_sqft  bath   price  bhk
0   Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1           Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2                Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3         Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                   Kothanur      2 BHK      1200.0   2.0   51.00    2
5                 Whitefield      2 BHK      1170.0   2.0   38.00    2
6           Old Airport Road      4 BHK      2732.0   4.0  204.00    4
7               Rajaji Nagar      4 BHK      3300.0   4.0  600.00    4
8               Marathahalli      3 BHK      1310.0   3.0   63.25    3
9               Gandhi Bazar  6 Bedroom      1020.0   6.0  370.00    6
10                Whitefield      3 BHK      1800.0   2.0   70.00    3
11                Whitefield  4 Bedroom      2785.0   5.0  295.00    4
12        7th Phase JP Nagar      2 BHK      1000.0   2.0   38.00    2
13                 Gottigere      2 BHK      1100.0   2.0   40.00    2
14                  Sarjapur  3 Bedroom      2250.0   3.0  148.00    3
15               Mysore Road      2 BHK      1175.0   2.0   73.50    2
16             Bisuvanahalli      3 BHK      1180.0   3.0   48.00    3
17     Raja Rajeshwari Nagar      3 BHK      1540.0   3.0   60.00    3
18     Ramakrishnappa Layout      3 BHK      2770.0   4.0  290.00    3
19        Manayata Tech Park      2 BHK      1100.0   2.0   48.00    2
20                   Kengeri      1 BHK       600.0   1.0   15.00    1
21                Binny Pete      3 BHK      1755.0   3.0  122.00    3
22               Thanisandra  4 Bedroom      2800.0   5.0  380.00    4
23                 Bellandur      3 BHK      1767.0   3.0  103.00    3
24               Thanisandra       1 RK       510.0   1.0   25.25    1
25           Mangammanapalya      3 BHK      1250.0   3.0   56.00    3
26           Electronic City      2 BHK       660.0   1.0   23.10    2
27                Whitefield      3 BHK      1610.0   3.0   81.00    3
28          Ramagondanahalli      2 BHK      1151.0   2.0   48.77    2
29           Electronic City      3 BHK      1025.0   2.0   47.00    3
30                 Yelahanka      4 BHK      2475.0   4.0  186.00    4

In [22]:
df4.loc[30]

location      Yelahanka
size              4 BHK
total_sqft       2475.0
bath                4.0
price             186.0
bhk                   4
Name: 30, dtype: object

In [23]:
df5 = df4.copy()

### Feature Engineering

In [24]:
df5['price_per_sqft'] = df5['price']*100000/df5['total_sqft']
df5.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0     3699.810606  
1     4615.384615  
2     4305.555556  
3     6245.890861  
4     4250.000000

In [25]:
df5['location'].unique()

array(['Electronic City Phase II', 'Chikka Tirupathi', 'Uttarahalli', ...,
       '12th cross srinivas nagar banshankari 3rd stage',
       'Havanur extension', 'Abshot Layout'], dtype=object)

In [26]:
len(df5['location'].unique())

1304

##### Classic High Dimensionality problem. Dimensionality Curse. Cant do One hot encoding, coz it will end up creating 1304 extra columns. We can use other method. for ex here locations with less than 10 data pts can be converted to "Other Locations"

In [27]:
df5.location = df5.location.apply(lambda x: x.strip())

In [28]:
location_stats = df5.groupby('location')['location'].agg('count').sort_values(ascending=False)
location_stats

location
Whitefield               535
Sarjapur  Road           392
Electronic City          304
Kanakpura Road           266
Thanisandra              236
                        ... 
1 Giri Nagar               1
Kanakapura Road,           1
Kanakapura main  Road      1
Karnataka Shabarimala      1
whitefiled                 1
Name: location, Length: 1293, dtype: int64

In [29]:
len(location_stats[location_stats<=10])

1052

In [30]:
location_stats_less_than_10 = location_stats[location_stats<=10]
location_stats_less_than_10

location
Basapura                 10
1st Block Koramangala    10
Gunjur Palya             10
Kalkere                  10
Sector 1 HSR Layout      10
                         ..
1 Giri Nagar              1
Kanakapura Road,          1
Kanakapura main  Road     1
Karnataka Shabarimala     1
whitefiled                1
Name: location, Length: 1052, dtype: int64

In [31]:
len(df5.location.unique())

1293

In [32]:
# performing a data transformation
df5.location = df5.location.apply(lambda x: "other" if x in location_stats_less_than_10 else x)
len(df5.location.unique())

242

In [34]:
df5.head(10)

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   
5                Whitefield      2 BHK      1170.0   2.0   38.00    2   
6          Old Airport Road      4 BHK      2732.0   4.0  204.00    4   
7              Rajaji Nagar      4 BHK      3300.0   4.0  600.00    4   
8              Marathahalli      3 BHK      1310.0   3.0   63.25    3   
9                     other  6 Bedroom      1020.0   6.0  370.00    6   

   price_per_sqft  
0     3699.810606  
1     4615.384615  
2     4305.555556  
3     6245.890861  
4     4250.000000  
5     3247.863248  
6     7467.057101  
7    18181.818182  
8     4828.244275  
9    36274.509804

### Outlier Removal

In [40]:
df5[df5['total_sqft']/df5['bhk']<300].head()

location       size  total_sqft  bath  price  bhk  \
9                 other  6 Bedroom      1020.0   6.0  370.0    6   
45           HSR Layout  8 Bedroom       600.0   9.0  200.0    8   
58        Murugeshpalya  6 Bedroom      1407.0   4.0  150.0    6   
68  Devarachikkanahalli  8 Bedroom      1350.0   7.0   85.0    8   
70                other  3 Bedroom       500.0   3.0  100.0    3   

    price_per_sqft  
9     36274.509804  
45    33333.333333  
58    10660.980810  
68     6296.296296  
70    20000.000000

In [41]:
df5.shape

(13246, 7)

In [42]:
df6 = df5[~(df5['total_sqft']/df5['bhk']<300)]
df6.shape

(12502, 7)

In [43]:
df6.price_per_sqft.describe()

count     12456.000000
mean       6308.502826
std        4168.127339
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_per_sqft, dtype: float64

In [46]:
def remove_pps_outliers(df):
    df_out = pd.DataFrame()
    for key, subdf in df.groupby('location'):
        m = np.mean(subdf.price_per_sqft)
        st = np.std(subdf.price_per_sqft)
        reduced_df = subdf[(subdf.price_per_sqft>(m-st)) & (subdf.price_per_sqft<=(m+st))]
        df_out = pd.concat([df_out,reduced_df], ignore_index=True)
    return df_out

In [49]:
df7 = remove_pps_outliers(df6)
df7.shape

(10241, 7)